### PR TITLE
nixos/installer: add kexec-tarball

### PR DESCRIPTION
**Pull request by jfroche**
Create a tarball including kexec and all the required nix dependencies.
Enable to use kexec on system where kexec is not available and/or where you don't want to depend on existing kexec.

This code is a copy code from nix-community/nixos-generators and cleverca22/nix-tests 

cc @cleverca22 
cc @Mic92

### CODE REVIEW

**Comment by roberth** [comment]
This is very similar to https://github.com/NixOS/nixpkgs/pull/161253
Does the existing module need to be improved? (It certainly needs docs, and I should have asked for those; sorry)

**Comment by jfroche** [comment]
kexec-boot.nix relies on the fact that kexec is installed on the host. This function create a tartball ready to be used on any Linux distro without relying on external software (appart from tar).
We use it to bootstrap nixos on physical host  like hetzner booting on minimal rescue Linux distribution. The code is already on
nixos-generate https://github.com/nix-community/nixos-generators/blob/master/formats/kexec.nix which rely on https://github.com/cleverca22/nix-tests/blob/master/kexec/kexec.nix . system.build.image in the later is similar to what kexec-boot provides.
Do you think kexec.nix should be merged with kexec-boot.nix ?
@flokli as you are the original author of kexec-boot you might have an opinion on this ?

**Comment by flokli** [comment]
`kexec-boot.nix` from https://github.com/NixOS/nixpkgs/pull/161253 exposes a `config.system.build.kexecBoot` attribute, returning a directory with 3 files: the new kernel image, initrd and a shellscript invoking `kexec` with these files (assumed to be in the same directory).

These three files are meant to be copied to the other host. It explicitly doesn't bring bash, the kexec binary or anything else.

The kexec-tarball introduced here brings the `kexec` binary and all of its transitive dependencies in `nix/store` inside the tarball, which needs to be extracted to `/`. This assumes that location is something the user is able to write to.

Even if that was fine, or we'd have a static version of the binary that's more portable, we shouldn't try to "bringing kexec with us" - it requires support in the currently running kernel, and at least Debian has two ways to plumb it into the system. This should be the realm of the package manager of the distro (still) running on the machine.

**Comment by Mic92** [comment]
I have not saw any incompatibilities issues with delivering an own `kexec` in various Linux distributions. Can you elaborate on that? Afaik it's a kernel api that should be the same everywhere. kexec tools might be not as clean as doing `systemctl kexec`, but if one does not care about the previous system than the old state does not really matter.

> The kexec-tarball introduced here brings the kexec binary and all of its transitive dependencies in nix/store inside the tarball, which needs to be extracted to /. This assumes that location is something the user is able to write to.

You need root permission to do kexec anyway, so writing permissions is not a limitation. The transitive dependencies is actually whole point of the tarball. Extracting to `/` is fine if you don't intend to keep the original OS, which is the main use case for kexec-based installers as far as I cann see..

**Comment by flokli** [comment]
> I have not seen any incompatibilities issues with delivering an own kexec in various Linux distributions I tested this kexec tarball before. Can you elaborate on that?

Alpine Linux' kernel doesn't ship kexec functionality. Some distros might set `LD_LIBRARY_PATH` or other variables, that might leak in incompatible linkers.

> You need root permission to do kexec anyway, so writing permissions can be expected. […] Extracting to / is fine if you don't intend to keep the original OS

In some cases you still can't. NixOS mounts /nix/store read-only, even for root, and some other "immutable" distros (FCOS maybe?) also may disallow creating new folders in `/`.


TL;DR; there's plenty of reasons why the `kexec tarball` doesn't work, and I don't too much of a reason why we can't tell people to install `kexec-tools` via their distro, and then use the script.

**Comment by roberth** [comment]
I do think NixOS should at least merge the extra functionality into its kexec feature, like copying the ssh keys.

> kexec-boot.nix relies on the fact that kexec is installed on the host

> This assumes that location is something the user is able to write to.

> Some distros might set LD_LIBRARY_PATH or other variables, that might leak in incompatible linkers.

All solvable by including a static build of `kexec-tools` and perhaps `cpio`. These static builds could be added to the end of PATH and a boolean NixOS option could make the inclusion of tools configurable to save some bytes when you know that the old distro provides them.

**Comment by Mic92** [comment]
This is what our kexec nixos-installer does
